mlrun/datastore/datastore_profile.py

```diff
@@ -0,0 +1,160 @@
+
+    def url_with_credentials(self):
+        parsed_url = urlparse(self.endpoint_url)
+        username = self.username.get() if self.username else None
+        password = self.password.get() if self.password else None
+        if username:
+            if password:
+                netloc = f"{username}:{password}@{parsed_url.hostname}"
+            else:
+                netloc = f"{username}@{parsed_url.hostname}"
+
+        if parsed_url.port:
+            netloc += f":{parsed_url.port}"
+
+        new_parsed_url = ParseResult(
+            scheme=parsed_url.scheme,
+            netloc=netloc,
+            path=parsed_url.path,
+            params=parsed_url.params,
+            query=parsed_url.query,
+            fragment=parsed_url.fragment,
+        )
+        return urlunparse(new_parsed_url)
+
+
+class DatastoreProfile2Json(pydantic.BaseModel):
+    @staticmethod
+    def _to_json(attributes):
+        # First, base64 encode the values
+        encoded_dict = {
+            k: base64.b64encode(str(v).encode()).decode() for k, v in attributes.items()
+        }
+        # Then, return the dictionary as a JSON string with no spaces
+        return json.dumps(encoded_dict).replace(" ", "")
+
+    @staticmethod
+    def get_json_public(profile: DatastoreProfile) -> str:
+        return DatastoreProfile2Json._to_json(
+            {k: v for k, v in profile.dict().items() if not isinstance(v, dict)}
+        )
+
+    @staticmethod
+    def get_json_private(profile: DatastoreProfile) -> str:
+        return DatastoreProfile2Json._to_json(
+            {k: v for k, v in profile.dict().items() if isinstance(v, dict)}
+        )
+
+    @staticmethod
+    def create_from_json(public_json: str, private_json: str = "{}"):
+        attr1 = json.loads(public_json)
+        attr2 = json.loads(private_json)
+        attributes = {**attr1, **attr2}
+        decoded_dict = {
+            k: base64.b64decode(str(v).encode()).decode() for k, v in attributes.items()
+        }
+        datastore_type = decoded_dict.get("type")
+        if datastore_type == "redis":
+            return DatastoreProfileRedis.parse_obj(decoded_dict)
+        else:
+            if datastore_type:
+                reason = f"unexpected type '{decoded_dict['type']}'"
+            else:
+                reason = "missing type"
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                f"Datastore profile failed to create from json due to {reason}"
+            )
+
+
+def datastore_profile_read(url):
+    parsed_url = urlparse(url)
+    if parsed_url.scheme.lower() != "ds":
+        raise mlrun.errors.MLRunInvalidArgumentError(
+            f"resource URL '{url}' cannot be read as a datastore profile because its scheme is not 'ds'"
+        )
+
+    profile_name = parsed_url.hostname
+    project_name = parsed_url.username or mlrun.mlconf.default_project
+    public_profile = mlrun.db.get_run_db().get_datastore_profile(
```

Comment:
`get_datastore_profile` may return `None` for non-existent profile, right? Should handle this here.